runtime/src/iree/hal/drivers/vulkan/native_executable.cc

```diff
@@ -230,88 +222,657 @@ static iree_status_t iree_hal_vulkan_executable_flatbuffer_verify(
+  VkShaderModule* shader_modules = NULL;
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_allocator_malloc(logical_device->host_allocator(),
+                                shader_module_count * sizeof(shader_modules[0]),
+                                (void**)&shader_modules));
+
+  iree_status_t status = iree_ok_status();
+  for (iree_host_size_t i = 0; i < shader_module_count; ++i) {
+    iree_hal_vulkan_ShaderModuleDef_table_t shader_module_def =
         iree_hal_vulkan_ShaderModuleDef_vec_at(shader_modules_vec, i);
-    size_t code_size = flatbuffers_uint32_vec_len(
-        iree_hal_vulkan_ShaderModuleDef_spirv_code_get(shader_module));
-    if (code_size == 0) {
-      return iree_make_status(
-          IREE_STATUS_INVALID_ARGUMENT,
-          "executable SPIR-V code in shader module #%zu is missing", i);
+    status = iree_hal_vulkan_create_shader_module(
+        logical_device, shader_module_def, &shader_modules[i]);
+    if (!iree_status_is_ok(status)) {
+      status = iree_status_annotate_f(status, "shader_modules[%" PRIhsz "]", i);
+      break;
     }
   }
 
-  flatbuffers_uint32_vec_t shader_module_indices_vec =
-      iree_hal_vulkan_ExecutableDef_shader_module_indices_get(executable_def);
-  size_t shader_module_index_count =
-      flatbuffers_vec_len(shader_module_indices_vec);
-  if (shader_module_index_count != expected_entry_point_count) {
-    return iree_make_status(
-        IREE_STATUS_INVALID_ARGUMENT,
-        "executable has %" PRIhsz
-        " entry points but %zu shader module indices are defined",
-        expected_entry_point_count, shader_module_index_count);
+  if (iree_status_is_ok(status)) {
+    *out_shader_module_count = shader_module_count;
+    *out_shader_modules = shader_modules;
+  } else {
+    iree_hal_vulkan_release_shader_modules(logical_device, shader_module_count,
+                                           shader_modules);
   }
-  for (size_t i = 0; i < shader_module_index_count; ++i) {
-    uint32_t index = flatbuffers_uint32_vec_at(shader_module_indices_vec, i);
-    if (index >= shader_module_count) {
-      return iree_make_status(
-          IREE_STATUS_INVALID_ARGUMENT,
-          "executable entry point shader module index %u out of range; "
-          "executable only has %zu total shader modules",
-          index, shader_module_count);
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+//===----------------------------------------------------------------------===//
+// Pipelines
+//===----------------------------------------------------------------------===//
+
+// Creates a pipeline from the set of available pipeline layouts and shader
+// modules and stores it into |out_pipeline|.
+//
+// NOTE: vkCreateComputePipelines takes multiple pipelines but doesn't speed up
+// creation on any known driver; we process one at a time so that we can get
+// better error messages and multithread the pipeline creation ourselves.
```

Comment:
Interesting - did you run those speed tests / benchmarks yourself, or did other people report that somewhere? (I'll buy the point about error messages, but startup time is pretty important for us, so would prefer to have high confidence that what we're doing is optimal)

Reply:
Yeah, drivers are not required and don't multithread pipeline creation internally (as a general rule Vulkan drivers don't maintain their own thread pools in user mode). The only savings is on the call into the driver which is tiny compared to actual creation and shader JIT that usually happens.

From the Vulkan guide:
> For compiling pipelines, vkCreateShaderModule and vkCreateGraphicsPipeline are both allowed to be called from multiple threads at once. A common approach for multithreaded shader compilation is to have a background thread dedicated to it, with it constantly looking into a parallel queue to receive compilation requests, and putting the compiled pipelines into another queue that then the main renderthread will connect to the simulation. This is very important to do if you want to have an engine that doesn’t have a lot of hitching. Compiling shader pipelines can take a very long time, so if you have to compile pipelines at runtime outside of a load screen, then you need to implement such a multithreaded async compile scheme for your game to work well.